Fix integration tests when running on the emulator

jskeet opened the pull request:
Please review one commit at a time.

The last of these commits is the most impactful one (in that it moves from "half the tests fail" to "everything passes") but is the only one I'm worried about in terms of production code. I'm happy for it to be merged for now, and even for betas to be released with it in, but we should think really carefully before including it in a GA release.

jskeet commented:
Whoops - will fix the broken unit test.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/EphemeralTransaction.cs:
```diff
@@ -95,22 +89,20 @@ async Task<IEnumerable<long>> Impl()
+                int count = await ((ISpannerTransaction)transaction)
+                    .ExecuteMutationsAsync(mutations, cancellationToken, timeoutSeconds)
+                    .ConfigureAwait(false);
+                rollbackCheckpoint();
```
Should this be after the commit? What state is the transaction left in if the commit fails? And it doesn't hurt us to put this after the commit anyway (I think). And same for the others except dml which is of its own.

jskeet replied:
No, it should be before the commit - because I believe if *commit* fails, then it's automatically rolled back and we don't need to call rollback. Although I suppose there could be a communication problem with means the reason the commit fails is that the server never gets the message.

I'll leave it for now, and we'll see what the Spanner team thinks we should do.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/EphemeralTransaction.cs:
```diff
@@ -38,48 +39,41 @@ internal EphemeralTransaction(SpannerConnection connection, TransactionOptions t
+            async Task<long> Impl(SpannerTransaction transaction, Action rollbackCheckpoint)
             {
-                using (var transaction = await _connection.BeginTransactionImplAsync(_transactionOptions, TransactionMode.ReadWrite, cancellationToken).ConfigureAwait(false))
+                // TODO: Validate this. Do we really never need to roll back the transaction here?
```
Might depend on partioned vs. non-partioned?

jskeet replied:
Maybe. It's frankly unclear to me.

skuruppu replied:
Oooh this is something I'm not confident to comment on. I will have to defer to @snehashah16, especially given that this is something that the emulator is expecting but is not an issue on prod.

amanda-tarafa commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/EphemeralTransaction.cs:
```diff
@@ -127,5 +119,45 @@ async Task<ReliableStreamReader> Impl()
+                    }
+                    finally
+                    {
+                        if (rollback)
```
Why not simply catch any exception and _attempt_ a rollback? Is it because of partioned-dml? What happens if we rollback that transaction? I'd say it would rollback to the last commit which is fine?

jskeet replied:
It's because I believe we shouldn't call Rollback if we've already called Commit - that a failed Commit already rolls back.
But let's wait to hear more.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/AssemblyInfo.cs:
```diff
@@ -11,9 +11,10 @@
 // Uncomment these to run the tests in a predictable order. (By fixture name, then by test case within the fixture.)
-// [assembly: TestCollectionOrderer(AlphabeticalOrderer.TypeName, AlphabeticalOrderer.AssemblyName)]
-// [assembly: TestCaseOrderer(AlphabeticalOrderer.TypeName, AlphabeticalOrderer.AssemblyName)]
+[assembly: TestCollectionOrderer(AlphabeticalOrderer.TypeName, AlphabeticalOrderer.AssemblyName)]
```
Okay. We can just remove the commit.

jskeet commented:
Hmm... I'm now seeing the session label test failing - it's finding 160-165 sessions with the specified label instead of 100. I need to dig into that.

jskeet commented:
@skuruppu: It looks like the emulator isn't actually filtering as requested on session labels. Do you know anything about this as a limitation? (I'm surprised the tests didn't fail earlier in the day...)

jskeet commented:
Okay, the session label filtering is a known limitation of the emulator. I'm skipping that test. I've also fixed a bad test which was trying to use "creating a database with a name of BadName" as an example of DDL that would fail. It fails *after the first success per instance* but that's not really good enough. Have changed it to use Bad-Name instead, which is invalid due to using a hyphen.

skuruppu commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/PartitionedReadTests.cs:
```diff
@@ -29,9 +29,10 @@ public class PartitionedReadTests
+        [SkippableFact]
         public async Task DistributedReadAsync()
         {
+            Skip.If(_fixture.RunningOnEmulator, "Emulator doesn't support distributed reads yet.");
```
Just as an FYI, this test does pass if you use the daily build of the emulator but I gather that this is not available through gcloud yet.

jskeet commented:
Okay, I've now:

- Removed the previous "last commit" that changed EphemeralTransaction
- Added another commit to skip BatchDmlTest.InvalidDmlBatch_PartialSuccess_ExplicitTransaction, which leaks a transaction
- Raised #5021 to consider the best fix for leaking transactions
- Added a final commit to skip a Snippet test (so the snippet suite now passes on the emulator too)

The only changes in production code are now:

- Working around the emulator returning a ExecuteBatchDmlResponse without a status, in SpannerTransaction
- Reverting the previous SessionPoolManager change to use special options for the emulator

jskeet commented:
Will merge on green - thanks!
